batch/test/test_batch.py

```diff
@@ -1128,8 +1121,88 @@ def test_update_batch_w_deps_in_update(client: BatchClient):
     status = j2.wait()
 
     assert status['state'] == 'Success', str((status, b.debug_info()))
-    assert j1.job_id == 2, str(b.debug_info())
-    assert j2.job_id == 3, str(b.debug_info())
+    assert b.status()['state'] == 'success', str((b.status(), b.debug_info()))
+    assert b.status()['n_jobs'] == 3, str(b.debug_info())
+
+
+def test_update_batch_w_deps_in_update_always_run(client: BatchClient):
+    bb = client.create_batch()
+    j = bb.create_job(DOCKER_ROOT_IMAGE, ['false'])
+    bb.submit()
+
+    j.wait()
+
+    j1 = bb.create_job(DOCKER_ROOT_IMAGE, ['true'], always_run=True)
+    j2 = bb.create_job(DOCKER_ROOT_IMAGE, ['true'], parents=[j, j1])
+    b = bb.submit()
+    status = j2.wait()
+
+    assert status['state'] == 'Success', str((status, b.debug_info()))
+    assert b.status()['state'] == 'failure', str((b.status(), b.debug_info()))
+
+
+def test_update_batch_w_failing_deps_in_update(client: BatchClient):
+    bb = client.create_batch()
+    j = bb.create_job(DOCKER_ROOT_IMAGE, ['true'])
+    bb.submit()
+
+    j1 = bb.create_job(DOCKER_ROOT_IMAGE, ['false'])
+    j2 = bb.create_job(DOCKER_ROOT_IMAGE, ['true'], parents=[j, j1])
+    b = bb.submit()
+    j2_status = j2.wait()
+
+    assert j.status()['state'] == 'Success', str((j.status(), b.debug_info()))
+    assert j2_status['state'] == 'Cancelled', str((j2_status, b.debug_info()))
+    assert b.status()['state'] == 'failure', str((b.status(), b.debug_info()))
+
+
+def test_update_with_always_run(client: BatchClient):
+    bb = client.create_batch()
+    j1 = bb.create_job(DOCKER_ROOT_IMAGE, ['sleep 3600'])
+    b = bb.submit()
+
+    j2 = bb.create_job(DOCKER_ROOT_IMAGE, ['true'], always_run=True, parents=[j1])
+    b = bb.submit()
+
+    delay = 0.1
+    while True:
+        if j1.is_complete():
+            assert False, str(j1.status(), b.debug_info())
+        if j1.is_running():
+            break
+        delay = await sync_sleep_and_backoff(delay)
+
+    assert j2.is_pending(), str(j2.status(), b.debug_info())
+
+    b.cancel()
+    j2.wait()
+
+    assert j2.status()['state'] == 'Success', str((j2.status(), b.debug_info()))
+    assert b.status()['state'] == 'cancelled', str((b.status(), b.debug_info()))
```

Comment:
This breaks the abstraction a bit here, so we would have to reach into the underlying aioclient methods, but I think it's worth it to have the following test as well:

1. Submit an hour long sleep job, `j1`
2. Create an update and submit a `true` job, `j2` that depends on `j1`, but don't commit the update yet
3. Cancel `j1`
4. Verify that `j2` is `Pending`
5. Commit the update
6. Verify that `j2` is `Cancelled`

Reply:
Unfortunately, I can't write this test for two reasons: 
1. we can't cancel a specific job -- only a batch.
2. we can't query a job that hasn't been committed 

FYI - I had to modify more queries to get the desired behavior.

Reply:
All good, I think the test that submits an update with a job dependent on one that has already failed is good enough to test that on commit we will cancel a job with failed parents.